csi.proto

```diff
@@ -1634,3 +1679,190 @@ message NodeExpandVolumeResponse {
   // The capacity of the volume in bytes. This field is OPTIONAL.
   int64 capacity_bytes = 1;
 }
+message GroupControllerGetCapabilitiesRequest {
+  // Intentionally empty.
+}
+
+message GroupControllerGetCapabilitiesResponse {
+  // All the capabilities that the group controller service supports.
+  // This field is OPTIONAL.
+  repeated GroupControllerServiceCapability capabilities = 1;
+}
+
+// Specifies a capability of the group controller service.
+message GroupControllerServiceCapability {
+  message RPC {
+    enum Type {
+      UNKNOWN = 0;
+
+      // Indicates that the group controller plugin supports
+      // creating and deleting a volume group snapshot.
+      CREATE_DELETE_VOLUME_GROUP_SNAPSHOT = 1
+      [(alpha_enum_value) = true];
+
+      // Indicates whether the group controller plugin supports
+      // creating a volume from an individual volume snapshot if the
+      // volume snapshot is part of a VolumeGroupSnapshot.
+      // Use cases: selective restore, advanced recovery, etc.
+      INDIVIDUAL_SNAPSHOT_RESTORE = 2 [(alpha_enum_value) = true];
+
+      // Indicates that the group controller plugin supports getting
+      // details of a volume group snapshot.
+      GET_VOLUME_GROUP_SNAPSHOT = 3 [(alpha_enum_value) = true];
+
+      // Indicates that the group controller plugin supports getting
+      // details of a list of volume group snapshots.
+      LIST_VOLUME_GROUP_SNAPSHOTS = 4 [(alpha_enum_value) = true];
+    }
+
+    Type type = 1;
+  }
+
+  oneof type {
+    // RPC that the controller supports.
+    RPC rpc = 1;
+  }
+}
+message CreateVolumeGroupSnapshotRequest {
+  option (alpha_message) = true;
+
+  // suggested name for a group snapshot (required for idempotent)
+  // This field is REQUIRED.
+  string name = 1;
+
+  // volume ids of the volumes to be snapshotted together.
+  // This field is REQUIRED.
+  repeated string volume_ids = 2;
+
+  // Secrets required by plugin to complete volume group snapshot
+  // creation request.
+  // This field is OPTIONAL. Refer to the `Secrets Requirements`
+  // section on how to use this field.
+  map<string, string> secrets = 3 [(.csi.v1.csi_secret) = true];
```

Comment:
Technically we can make this a list of maps, but that makes it very complicated. I wonder if we should require user to use one secret per group.

Reply:
My concern here is that this usage is different in kind than other usages of secrets in the interface. CSI accommodates the idea of per-volume and per-snapshot secrets, because the APIs that create/delete these objects never deal with more than one at a time (possibly both 1 volume + 1 snapshot at the same time).

List APIs are the only ones that deal with more than object of the same type at the same time, and we've seen serious problems implementing secret handling in those cases. Kubernetes sidesteps the issue by simply not using List APIs for anything other than single objects.

Here we won't be able to sidestep the issue. I worry that simply asserting "Secrets MUST be valid for creating group snapshot from all the volumes." only rules out the possibility of implementing this feature with per-volume secrets. Assuming that the CO has a bunch of secrets for the volumes in question and they're all different, what should it do? Use a different secret for snapshot creating? Pass the various volume secrets by multiplexing them into the map? We owe it to the CO to spell out what to do with this field so we don't end up in the mess we have for the List APIs where they're basically unusable if you have secrets.